File: 2021/day03/day03.py
#!/usr/bin/env python3
import sys

if len(sys.argv) > 1:
    filename = sys.argv[1]
else:
    filename = "input.txt"

data = open(filename).read().splitlines()

# star 1
counts = [[0, 0] for _ in range(len(data[0]))]
for i in data:
    for j, k in enumerate(i):
        counts[j][int(k)] += 1

gamma = int("".join("0" if i > j else "1" for i, j in counts), base=2)
epsilon = int("".join("1" if i > j else "0" for i, j in counts), base=2)
print(gamma * epsilon)

# star 2
def filter_by_bit(data, bit, negate):
    ones = 0
    for i in data:
        ones += int(i[bit])
    zeroes = len(data) - ones

    criteria = "1" if negate ^ (ones >= zeroes) else "0"
    return list(filter(lambda i: i[bit] == criteria, data))


def filter_until_one(data, negate):
    bit = 0
    while True:
        data = filter_by_bit(data, bit, negate)
        if len(data) <= 1:
            break
        bit += 1

    return data[0]


oxygen = int(filter_until_one(data, False), base=2)
co2 = int(filter_until_one(data, True), base=2)
print(oxygen * co2)

File: 2021/day03/day03.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script

import System.Environment (getArgs)
import Data.List (transpose)

countBits :: [Bool] -> (Int, Int)
countBits = go 0 0
  where go z o (True  : xs) = go z (o + 1) xs
        go z o (False : xs) = go (z + 1) o xs
        go z o []           = (z, o)

fromBinary :: [Bool] -> Int
fromBinary = go 0
  where go n (x:xs) = go (2 * n + fromEnum x) xs
        go n []     = n

filterByBit :: Bool -> Int -> [[Bool]] -> [[Bool]]
filterByBit negate n xs = filter ((== criteria) . (!! n)) xs
  where (zeroes, ones) = countBits $ map (!! n) xs
        criteria = negate /= (ones >= zeroes)

filterUntilOne :: Bool -> [[Bool]] -> [Bool]
filterUntilOne negate = go 0
  where go n xs@(_:_:_) = go (n + 1) $ filterByBit negate n xs
        go _ [x]        = x
        go _ []         = error "unreachable"

star1 :: [[Bool]] -> Int
star1 xs = fromBinary gamma * fromBinary epsilon
  where counts = map countBits $ transpose xs
        gamma = map (\(zeroes, ones) -> ones >= zeroes) counts
        epsilon = map not gamma

star2 :: [[Bool]] -> Int
star2 xs = fromBinary oxygen * fromBinary co2
  where oxygen = filterUntilOne True xs
        co2 = filterUntilOne False xs

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  input <- map (map (=='1')) . lines <$> readFile filename
  print $ star1 input
  print $ star2 input
